mm2src/coins/sia/http_client.rs

```diff
@@ -0,0 +1,189 @@
+use serde::de::DeserializeOwned;
+use std::ops::Deref;
+use std::sync::Arc;
+
+use mm2_number::MmNumber;
+
+/// HTTP(s) client for Sia-protocol coins
+#[derive(Debug)]
+pub struct SiaHttpClientImpl {
+    /// Name of coin the http client is intended to work with
+    pub coin_ticker: String,
+    /// The uri to send requests to
+    pub uri: String,
+    /// Value of Authorization header password, e.g. "Basic base64(:password)"
+    pub auth: String,
+}
+
+#[derive(Clone, Debug)]
+pub struct SiaApiClient(pub Arc<SiaApiClientImpl>);
+impl Deref for SiaApiClient {
+    type Target = SiaApiClientImpl;
+    fn deref(&self) -> &SiaApiClientImpl { &self.0 }
+}
+
+impl SiaApiClient {
+    pub fn new(_coin_ticker: &str, base_url: Url, auth: &str) -> Result<Self, SiaApiClientError> {
+        let new_arc = SiaApiClientImpl::new(base_url, auth)?;
+        Ok(SiaApiClient(Arc::new(new_arc)))
+    }
+}
+
+#[derive(Debug)]
+pub struct SiaApiClientImpl {
+    client: reqwest::Client,
+    base_url: Url,
+}
+
+// this is neccesary to show the URL in error messages returned to the user
+// this can be removed in favor of using ".with_url()" once reqwest is updated to v0.11.23
+#[derive(Debug)]
+pub struct ReqwestErrorWithUrl {
+    error: reqwest::Error,
+    url: Url,
+}
+
+impl Display for ReqwestErrorWithUrl {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Error: {}, URL: {}", self.error, self.url)
+    }
+}
+
+#[derive(Debug, Display)]
+pub enum SiaApiClientError {
+    Timeout(String),
+    BuildError(String),
+    ApiUnreachable(String),
+    ReqwestError(ReqwestErrorWithUrl),
+    UrlParse(url::ParseError),
+}
+
+impl From<SiaApiClientError> for String {
+    fn from(e: SiaApiClientError) -> Self { format!("{:?}", e) }
+}
+
+async fn fetch_and_parse<T: DeserializeOwned>(client: &Client, url: Url) -> Result<T, SiaApiClientError> {
+    client
+        .get(url.clone())
+        .send()
+        .await
+        .map_err(|e| {
+            SiaApiClientError::ReqwestError(ReqwestErrorWithUrl {
+                error: e,
+                url: url.clone(),
+            })
+        })?
+        .json::<T>()
+        .await
+        .map_err(|e| SiaApiClientError::ReqwestError(ReqwestErrorWithUrl { error: e, url }))
+}
+
+// https://github.com/SiaFoundation/core/blob/4e46803f702891e7a83a415b7fcd7543b13e715e/types/types.go#L181
+#[derive(Deserialize, Serialize, Debug)]
+pub struct GetConsensusTipResponse {
+    pub height: u64,
+    pub id: String, // TODO this can match "BlockID" type
+}
+
+// https://github.com/SiaFoundation/walletd/blob/9574e69ff0bf84de1235b68e78db2a41d5e27516/api/api.go#L36
+// https://github.com/SiaFoundation/walletd/blob/9574e69ff0bf84de1235b68e78db2a41d5e27516/wallet/wallet.go#L25
+#[derive(Deserialize, Serialize, Debug)]
+pub struct GetAddressesBalanceResponse {
+    pub siacoins: MmNumber,
+    #[serde(rename = "immatureSiacoins")]
+    pub immature_siacoins: MmNumber,
+    pub siafunds: u64
```

Comment:
I know that we are not going to implement swaps for siafunds, but are we gonna implement wallet operations for them?

Reply:
Undetermined at this point. It's not within the scope of the grant. The reference to siafunds here is due to the specific API endpoint this struct is being deserialized from. It will be ignored for the time being.